File: frontend/src/services/accountService.ts
import axiosInstance from './axios';

export interface AccountRegistration {
  id?: string;
  user_id: number;
  account_number: string;
  balance: number;
  account_type_id?: number;
  type_name?: string;
  Description?: string;
  currency?: string;
  is_active?: boolean;
  created_at?: string;
  updatedAt?: string;
  name?: string;
}

export interface AccountType {
  ID: number;
  type_name: string;
  Description: string;
  currency: string;
}

export interface AccountDetails {
  account_number: string;
  full_name: string;
}

export const accountService = {
  // Create new account registration
  async create(account: Omit<AccountRegistration, 'id' | 'status' | 'createdAt' | 'updatedAt'>) {
    try {
      console.log('Creating account with data:', account);
      const response = await axiosInstance.post('/api/user/accounts', account);
      console.log('Create account response:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Error creating account:', error.response?.data || error.message);
      throw error;
    }
  },

  // Get all account registrations
  async getAll(): Promise<AccountRegistration[]> {
    const user = JSON.parse(localStorage.getItem('user') || '{}');
    const response = await axiosInstance.get(`/api/user/accounts/${user.id}`);
    return response.data;
  },

  // Get single account registration
  async getById(id: string) {
    try {
      const response = await axiosInstance.get(`/api/user/accounts/${id}`);
      return response.data;
    } catch (error: any) {
      console.error('Error fetching account:', error.response?.data || error.message);
      throw error;
    }
  },

  // Update account registration
  async update(id: string, account: Partial<AccountRegistration>) {
    try {
      const response = await axiosInstance.put(`/api/user/accounts/${id}`, account);
      return response.data;
    } catch (error: any) {
      console.error('Error updating account:', error.response?.data || error.message);
      throw error;
    }
  },

  // Delete account registration
  async delete(id: string) {
    try {
      const response = await axiosInstance.delete(`/api/user/accounts/${id}`);
      return response.data;
    } catch (error: any) {
      console.error('Error deleting account:', error.response?.data || error.message);
      throw error;
    }
  },

  // Get all account types
  async getAccountTypes() {
    try {
      const response = await axiosInstance.get('/api/user/account-types');
      return response.data;
    } catch (error: any) {
      console.error('Error fetching account types:', error.response?.data || error.message);
      throw error;
    }
  },

  getAccountDetails: async (accountNumber: string): Promise<AccountDetails> => {
   try {
    const response = await axiosInstance.get(`/api/user/account-details?account_number=${accountNumber}`);
    return response.data;
   } catch (error: any) {
    console.error('Error fetching account details:', error.response?.data || error.message);
    throw error;
   }
  },
}; 

File: backend/services/account_service.go
package services

import (
	"bank/db"
	"bank/dtos"
	"bank/models"
	"errors"

	"github.com/lib/pq"
)


func CreateAccount(acc *models.Account) error {
	query := `INSERT INTO accounts (account_number, balance, user_id, account_type_id, created_at)
	          VALUES ($1, $2, $3, $4, NOW()) RETURNING id`
	err := db.GetDB().QueryRow(query, acc.AccountNumber, acc.Balance, acc.UserID, acc.AccountTypeID).
		Scan(&acc.ID)
	if err != nil {
		// Check for PostgreSQL unique constraint violation
		if pgErr, ok := err.(*pq.Error); ok && pgErr.Code == "23505" {
			if pgErr.Constraint == "accounts_account_number_key" {
				return errors.New("account number already exists")
			}
			return errors.New("duplicate value violates a unique constraint")
		}
		return err
	}

	// Log audit for create
	_ = LogAudit(&acc.UserID, "CREATE", "accounts", acc.ID, "Account created")
	return nil
}


// Get balance for a specific account by account ID
func GetAccountBalance(id string) (float64, error) {
	var balance float64
	query := `SELECT balance FROM accounts WHERE account_number = $1`
	err := db.GetDB().QueryRow(query, id).Scan(&balance)
	if err != nil {
		return 0, err
	}
	return balance, nil
}
// Get all accounts with their account type names (simulating Preload)
func GetAllAccounts() ([]models.Account, error) {
	query := `SELECT a.id, a.account_number, a.balance, a.user_id, a.account_type_id,  at.type_name AS account_type_name,at.description, currency,a.created_at
	          FROM accounts a
	          JOIN account_types at ON a.account_type_id = at.id`

	rows, err := db.DB.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var accounts []models.Account
	for rows.Next() {
		var acc models.Account
		var accountTypeName string
		var description string
		var currency string

		err := rows.Scan(&acc.ID, &acc.AccountNumber, &acc.Balance, &acc.UserID, &acc.AccountTypeID,
			&accountTypeName,&description,&currency, &acc.CreatedAt)
		if err != nil {
			return nil, err
		}

		acc.AccountType = models.AccountType{TypeName: accountTypeName,Description: description, Currency: currency}
		accounts = append(accounts, acc)
	}
	return accounts, nil
}

// Get a single account by ID (with account type)
func GetAccountsByUserID(userID uint) ([]dtos.AccountResponse, error) {
	query := `SELECT a.id, a.account_number, a.balance, a.user_id, a.account_type_id,
	                 at.type_name, at.description, at.currency,
	                 u.full_name, a.created_at
	          FROM accounts a
	          JOIN account_types at ON a.account_type_id = at.id
	          JOIN users u ON a.user_id = u.id
	          WHERE a.user_id = $1`

	rows, err := db.DB.Query(query, userID)
	if err != nil {
		return []dtos.AccountResponse{}, err // return empty slice instead of nil
	}
	defer rows.Close()

	responses := make([]dtos.AccountResponse, 0) // Always create a slice

	for rows.Next() {
		var acc dtos.AccountResponse
		err := rows.Scan(
			&acc.ID, &acc.AccountNumber, &acc.Balance, &acc.UserID, &acc.AccountTypeID,
			&acc.TypeName, &acc.Description, &acc.Currency,
			&acc.Name, &acc.CreatedAt,
		)
		if err != nil {
			return []dtos.AccountResponse{}, err
		}
		responses = append(responses, acc)
	}

	// Optionally check if no rows were found
	if len(responses) == 0 {
		return responses, nil // or custom error/info message if needed
	}

	return responses, nil
}




// Update an account
func UpdateAccount(id uint, updated *models.Account) error {
	query := `UPDATE accounts 
	          SET account_number = $1, balance = $2, user_id = $3, account_type_id = $4
	          WHERE id = $5`
	result, err := db.GetDB().Exec(query, updated.AccountNumber, updated.Balance, updated.UserID, updated.AccountTypeID, id)
	if err != nil {
		return err
	}
	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		return errors.New("no record updated")
	}

	// Log audit for update
	_ = LogAudit(&updated.UserID, "UPDATE", "accounts", id, "Account updated")
	return nil
}


// Delete an account
func DeleteAccount(id uint) error {
	query := `DELETE FROM accounts WHERE id = $1`
	result, err := db.GetDB().Exec(query, id)
	if err != nil {
		return err
	}
	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		return errors.New("no record deleted")
	}

	// Log audit for delete 
	_ = LogAudit(nil, "DELETE", "accounts", id, "Account deleted")
	return nil
}


func GetUserByAccountNumber( accountNumber string) (*dtos.AccoutResponse, error) {
    var user dtos.AccoutResponse
    query := `SELECT u.full_name, a.account_number
        FROM users u
        JOIN accounts a ON u.id = a.user_id
		where a.account_number = $1
        LIMIT 1`
    err :=db.GetDB().QueryRow(query, accountNumber).Scan( &user.FullName,&user.AccountNumber,)
    if err != nil {
        return nil, err
    }
    return &user, nil
}